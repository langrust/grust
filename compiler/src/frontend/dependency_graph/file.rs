use petgraph::algo::toposort;
use petgraph::graphmap::DiGraphMap;

prelude! {
    error::{Error, TerminationError},
    hir::file::File,
    symbol_table::SymbolTable,

}

impl File {
    /// Generate dependency graph for every nodes/component.
    pub fn generate_dependency_graphs(
        &mut self,
        symbol_table: &SymbolTable,
        errors: &mut Vec<Error>,
    ) -> Result<(), TerminationError> {
        let File { nodes, .. } = self;

        // initialize dictionariy for reduced graphs
        let mut nodes_reduced_graphs = HashMap::new();

        // create graph of nodes
        let mut nodes_graph = DiGraphMap::new();
        nodes
            .iter()
            .for_each(|node| node.add_node_dependencies(&mut nodes_graph));

        // sort nodes according to their dependencies
        let sorted_nodes = toposort(&nodes_graph, None).map_err(|node| {
            let error = Error::NotCausalNode {
                node: symbol_table.get_name(node.node_id()).clone(),
                location: self.location.clone(),
            };
            errors.push(error);
            TerminationError
        })?;
        nodes.sort_by(|n1, n2| {
            let index1 = sorted_nodes
                .iter()
                .position(|id| *id == n1.id)
                .expect("should be in sorted list");
            let index2 = sorted_nodes
                .iter()
                .position(|id| *id == n2.id)
                .expect("should be in sorted list");

            Ord::cmp(&index2, &index1)
        });

        // ordered nodes complete their dependency graphs
        nodes
            .iter_mut()
            .map(|node| node.compute_dependencies(symbol_table, &mut nodes_reduced_graphs, errors))
            .collect::<Vec<Result<(), TerminationError>>>()
            .into_iter()
            .collect::<Result<(), TerminationError>>()?;

        Ok(())
    }
}
