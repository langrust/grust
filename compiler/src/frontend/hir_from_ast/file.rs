prelude! {
    ast::Ast,
    hir::interface::Interface,
}

use super::HIRFromAST;

impl HIRFromAST for Ast {
    type HIR = hir::File;

    fn hir_from_ast(
        self,
        symbol_table: &mut SymbolTable,
        errors: &mut Vec<Error>,
    ) -> TRes<Self::HIR> {
        // initialize symbol table with builtin operators
        symbol_table.initialize();

        // store elements in symbol table
        self.store(symbol_table, errors)?;

        let Ast { items } = self;

        let (typedefs, functions, components, imports, exports, services) = items.into_iter().fold(
            (vec![], vec![], vec![], vec![], vec![], vec![]),
            |(
                mut typedefs,
                mut functions,
                mut components,
                mut imports,
                mut exports,
                mut services,
            ),
             item| {
                match item {
                    ast::Item::Component(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Function(function) => {
                        functions.push(function.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Typedef(typedef) => {
                        typedefs.push(typedef.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Service(service) => {
                        services.push(service.hir_from_ast(symbol_table, errors))
                    }
                    ast::Item::Import(import) => imports.push(
                        import
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::Export(export) => exports.push(
                        export
                            .hir_from_ast(symbol_table, errors)
                            .map(|res| (symbol_table.get_fresh_id(), res)),
                    ),
                    ast::Item::ComponentImport(component) => {
                        components.push(component.hir_from_ast(symbol_table, errors))
                    }
                }
                (typedefs, functions, components, imports, exports, services)
            },
        );

        let interface = Interface {
            services: services.into_iter().collect::<TRes<Vec<_>>>()?,
            imports: imports.into_iter().collect::<TRes<_>>()?,
            exports: exports.into_iter().collect::<TRes<_>>()?,
        };

        Ok(hir::File {
            typedefs: typedefs.into_iter().collect::<TRes<Vec<_>>>()?,
            functions: functions.into_iter().collect::<TRes<Vec<_>>>()?,
            components: components.into_iter().collect::<TRes<Vec<_>>>()?,
            interface,
            location: Location::default(),
        })
    }
}
