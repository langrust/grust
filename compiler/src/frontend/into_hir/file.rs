prelude! {
    ast::Ast,
    hir::interface::Interface,
}

impl IntoHir<hir::ctx::Simple<'_>> for Ast {
    type Hir = hir::File;

    fn into_hir(self, ctxt: &mut hir::ctx::Simple) -> TRes<Self::Hir> {
        // initialize symbol table with builtin operators
        ctxt.syms.initialize();

        // store elements in symbol table
        self.store(ctxt.syms, ctxt.errors)?;

        let Ast { items } = self;

        let (typedefs, functions, components, imports, exports, services) = items.into_iter().fold(
            (vec![], vec![], vec![], vec![], vec![], vec![]),
            |(
                mut typedefs,
                mut functions,
                mut components,
                mut imports,
                mut exports,
                mut services,
            ),
             item| {
                match item {
                    ast::Item::Component(component) => components.push(component.into_hir(ctxt)),
                    ast::Item::Function(function) => functions.push(function.into_hir(ctxt)),
                    ast::Item::Typedef(typedef) => typedefs.push(typedef.into_hir(ctxt)),
                    ast::Item::Service(service) => services.push(service.into_hir(ctxt)),
                    ast::Item::Import(import) => imports.push(
                        import
                            .into_hir(ctxt)
                            .map(|res| (ctxt.syms.get_fresh_id(), res)),
                    ),
                    ast::Item::Export(export) => exports.push(
                        export
                            .into_hir(ctxt)
                            .map(|res| (ctxt.syms.get_fresh_id(), res)),
                    ),
                    ast::Item::ComponentImport(component) => {
                        components.push(component.into_hir(ctxt))
                    }
                }
                (typedefs, functions, components, imports, exports, services)
            },
        );

        let interface = Interface {
            services: services.into_iter().collect::<TRes<Vec<_>>>()?,
            imports: imports.into_iter().collect::<TRes<_>>()?,
            exports: exports.into_iter().collect::<TRes<_>>()?,
        };

        Ok(hir::File {
            typedefs: typedefs.into_iter().collect::<TRes<Vec<_>>>()?,
            functions: functions.into_iter().collect::<TRes<Vec<_>>>()?,
            components: components.into_iter().collect::<TRes<Vec<_>>>()?,
            interface,
            location: Location::default(),
        })
    }
}
