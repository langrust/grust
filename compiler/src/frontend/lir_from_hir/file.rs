use crate::{
    hir::file::File,
    lir::{item::Item, project::Project},
    symbol_table::SymbolTable,
};

use super::LIRFromHIR;

impl LIRFromHIR for File {
    type LIR = Project;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let File {
            typedefs,
            functions,
            nodes,
            interface,
            ..
        } = self;

        let mut typedefs = typedefs
            .into_iter()
            .map(|typedef| typedef.lir_from_hir(symbol_table))
            .collect();
        let mut functions = functions
            .into_iter()
            .map(|function| function.lir_from_hir(symbol_table))
            .map(Item::Function)
            .collect();
        let mut state_machines = nodes
            .into_iter()
            .map(|node| node.lir_from_hir(symbol_table))
            .map(Item::StateMachine)
            .collect();
        let execution_machine = interface.lir_from_hir(symbol_table);

        let mut items = vec![];
        items.append(&mut typedefs);
        items.append(&mut functions);
        items.append(&mut state_machines);
        items.push(Item::ExecutionMachine(execution_machine));

        Project { items }
    }
}
