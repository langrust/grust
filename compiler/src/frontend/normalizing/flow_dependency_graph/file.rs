prelude! { just
    hir::File,
}

impl File {
    /// Generate dependency graphs for the interface.
    #[inline]
    pub fn generate_flows_dependency_graphs(&mut self) {
        self.interface.generate_flows_dependency_graphs()
    }
}
