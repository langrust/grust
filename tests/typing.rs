use codespan_reporting::files::{Files, SimpleFiles};

use grustine::ast::file::File;
use grustine::error::display;
use grustine::parser::langrust;

#[test]
fn typing_counter() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/counter.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(counter_id, &files.source(counter_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn typing_blinking() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/blinking.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn typing_button_management() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/button_management.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn typing_button_management_condition_match() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string("tests/fixture/button_management_condition_match.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn typing_button_management_using_function() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string("tests/fixture/button_management_using_function.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(blinking_id, &files.source(blinking_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn typing_pid() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/pid.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(pid_id, &files.source(pid_id).unwrap())
        .unwrap();
    file.typing(&mut errors).unwrap();

    insta::assert_yaml_snapshot!(file);
}

#[test]
fn error_when_typing_counter_not_well_typed() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_not_well_typed_id = files.add(
        "counter_not_well_typed.gr",
        std::fs::read_to_string("tests/fixture/counter_not_well_typed.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            counter_not_well_typed_id,
            &files.source(counter_not_well_typed_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_counter_unknown_signal() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_unknown_signal_id = files.add(
        "counter_unknown_signal.gr",
        std::fs::read_to_string("tests/fixture/counter_unknown_signal.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            counter_unknown_signal_id,
            &files.source(counter_unknown_signal_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_blinking_unknown_node() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_unknown_node_id = files.add(
        "blinking_unknown_node.gr",
        std::fs::read_to_string("tests/fixture/blinking_unknown_node.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            blinking_unknown_node_id,
            &files.source(blinking_unknown_node_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_button_management_using_function_unknown_element() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let button_management_using_function_unknown_element_id = files.add(
        "button_management_using_function_unknown_element.gr",
        std::fs::read_to_string(
            "tests/fixture/button_management_using_function_unknown_element.gr",
        )
        .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            button_management_using_function_unknown_element_id,
            &files
                .source(button_management_using_function_unknown_element_id)
                .unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_button_management_unknown_type() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let button_management_unknown_type_id = files.add(
        "button_management_unknown_type.gr",
        std::fs::read_to_string("tests/fixture/button_management_unknown_type.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            button_management_unknown_type_id,
            &files.source(button_management_unknown_type_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_pid_unknown_field() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_unknown_field_id = files.add(
        "pid_unknown_field.gr",
        std::fs::read_to_string("tests/fixture/pid_unknown_field.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            pid_unknown_field_id,
            &files.source(pid_unknown_field_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_pid_missing_field() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_missing_field_id = files.add(
        "pid_missing_field.gr",
        std::fs::read_to_string("tests/fixture/pid_missing_field.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            pid_missing_field_id,
            &files.source(pid_missing_field_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_blinking_component_call() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_component_call_id = files.add(
        "blinking_component_call.gr",
        std::fs::read_to_string("tests/fixture/blinking_component_call.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            blinking_component_call_id,
            &files.source(blinking_component_call_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_blinking_already_defined_element() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_already_defined_element_id = files.add(
        "blinking_already_defined_element.gr",
        std::fs::read_to_string("tests/fixture/blinking_already_defined_element.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            blinking_already_defined_element_id,
            &files.source(blinking_already_defined_element_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_blinking_incompatible_type() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let blinking_incompatible_type_id = files.add(
        "blinking_incompatible_type.gr",
        std::fs::read_to_string("tests/fixture/blinking_incompatible_type.gr")
            .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            blinking_incompatible_type_id,
            &files.source(blinking_incompatible_type_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_pid_incompatible_pattern() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_incompatible_pattern_id = files.add(
        "pid_incompatible_pattern.gr",
        std::fs::read_to_string("tests/fixture/pid_incompatible_pattern.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            pid_incompatible_pattern_id,
            &files.source(pid_incompatible_pattern_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_button_management_using_function_incompatible_input_number() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let button_management_using_function_incompatible_input_number_id = files.add(
        "button_management_using_function_incompatible_input_number.gr",
        std::fs::read_to_string(
            "tests/fixture/button_management_using_function_incompatible_input_number.gr",
        )
        .expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            button_management_using_function_incompatible_input_number_id,
            &files
                .source(button_management_using_function_incompatible_input_number_id)
                .unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_counter_expect_number() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_expect_number_id = files.add(
        "counter_expect_number.gr",
        std::fs::read_to_string("tests/fixture/counter_expect_number.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            counter_expect_number_id,
            &files.source(counter_expect_number_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_counter_expect_option() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let counter_expect_option_id = files.add(
        "counter_expect_option.gr",
        std::fs::read_to_string("tests/fixture/counter_expect_option.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            counter_expect_option_id,
            &files.source(counter_expect_option_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_pid_expect_structure_type() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_expect_structure_type_id = files.add(
        "pid_expect_structure_type.gr",
        std::fs::read_to_string("tests/fixture/pid_expect_structure_type.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            pid_expect_structure_type_id,
            &files.source(pid_expect_structure_type_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}

#[test]
fn error_when_typing_pid_unknown_enumeration() {
    let mut files = SimpleFiles::new();
    let mut errors = vec![];

    let pid_unknown_enumeration_id = files.add(
        "pid_unknown_enumeration.gr",
        std::fs::read_to_string("tests/fixture/pid_unknown_enumeration.gr").expect("unkown file"),
    );

    let mut file: File = langrust::fileParser::new()
        .parse(
            pid_unknown_enumeration_id,
            &files.source(pid_unknown_enumeration_id).unwrap(),
        )
        .unwrap();
    file.typing(&mut errors).unwrap_err();

    display(&errors, &files);
}
