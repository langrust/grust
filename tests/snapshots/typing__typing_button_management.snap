---
source: tests/typing.rs
expression: file
---
typedefs:
  - id: 19
    kind:
      Enumeration:
        elements:
          - 17
          - 18
    location: ~
  - id: 22
    kind:
      Enumeration:
        elements:
          - 20
          - 21
    location: ~
functions: []
nodes:
  - id: 26
    unscheduled_equations:
      25:
        id: 25
        expression:
          kind:
            Expression:
              expression:
                Application:
                  function_expression:
                    kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 15
                    typing:
                      Abstract:
                        - - Boolean
                          - Integer
                          - Integer
                        - Integer
                    location: ~
                    dependencies: ~
                  inputs:
                    - kind:
                        Expression:
                          expression:
                            Identifier:
                              id: 23
                      typing: Boolean
                      location: ~
                      dependencies: ~
                    - kind:
                        Expression:
                          expression:
                            Constant:
                              constant:
                                Integer: 0
                      typing: Integer
                      location: ~
                      dependencies: ~
                    - kind:
                        Expression:
                          expression:
                            Application:
                              function_expression:
                                kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 5
                                typing:
                                  Abstract:
                                    - - Integer
                                      - Integer
                                    - Integer
                                location: ~
                                dependencies: ~
                              inputs:
                                - kind:
                                    Expression:
                                      expression:
                                        Application:
                                          function_expression:
                                            kind:
                                              Expression:
                                                expression:
                                                  Identifier:
                                                    id: 2
                                            typing:
                                              Abstract:
                                                - - Integer
                                                - Integer
                                            location: ~
                                            dependencies: ~
                                          inputs:
                                            - kind:
                                                FollowedBy:
                                                  constant:
                                                    kind:
                                                      Expression:
                                                        expression:
                                                          Constant:
                                                            constant:
                                                              Integer: 0
                                                    typing: Integer
                                                    location: ~
                                                    dependencies: ~
                                                  expression:
                                                    kind:
                                                      Expression:
                                                        expression:
                                                          Identifier:
                                                            id: 25
                                                    typing: Integer
                                                    location: ~
                                                    dependencies: ~
                                              typing: Integer
                                              location: ~
                                              dependencies: ~
                                  typing: Integer
                                  location: ~
                                  dependencies: ~
                                - kind:
                                    Expression:
                                      expression:
                                        Identifier:
                                          id: 24
                                  typing: Integer
                                  location: ~
                                  dependencies: ~
                      typing: Integer
                      location: ~
                      dependencies: ~
          typing: Integer
          location: ~
          dependencies: ~
        location: ~
    unitary_nodes: {}
    contract:
      requires: []
      ensures: []
      invariant: []
    location: ~
    graph: ~
  - id: 35
    unscheduled_equations:
      31:
        id: 31
        expression:
          kind:
            Expression:
              expression:
                Match:
                  expression:
                    kind:
                      Expression:
                        expression:
                          Application:
                            function_expression:
                              kind:
                                Expression:
                                  expression:
                                    Identifier:
                                      id: 2
                              typing:
                                Abstract:
                                  - - Enumeration: Button
                                  - Enumeration: Button
                              location: ~
                              dependencies: ~
                            inputs:
                              - kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 27
                                typing:
                                  Enumeration: Button
                                location: ~
                                dependencies: ~
                    typing:
                      Enumeration: Button
                    location: ~
                    dependencies: ~
                  arms:
                    - - kind:
                          Enumeration:
                            enum_id: 19
                            elem_id: 17
                        typing:
                          Enumeration: Button
                        location: ~
                      - ~
                      - []
                      - kind:
                          Expression:
                            expression:
                              Enumeration:
                                enum_id: 22
                                elem_id: 21
                        typing:
                          Enumeration: ResetState
                        location: ~
                        dependencies: ~
                    - - kind:
                          Enumeration:
                            enum_id: 19
                            elem_id: 18
                        typing:
                          Enumeration: Button
                        location: ~
                      - ~
                      - []
                      - kind:
                          Expression:
                            expression:
                              Application:
                                function_expression:
                                  kind:
                                    Expression:
                                      expression:
                                        Identifier:
                                          id: 15
                                  typing:
                                    Abstract:
                                      - - Boolean
                                        - Enumeration: ResetState
                                        - Enumeration: ResetState
                                      - Enumeration: ResetState
                                  location: ~
                                  dependencies: ~
                                inputs:
                                  - kind:
                                      Expression:
                                        expression:
                                          Application:
                                            function_expression:
                                              kind:
                                                Expression:
                                                  expression:
                                                    Identifier:
                                                      id: 11
                                              typing:
                                                Abstract:
                                                  - - Integer
                                                    - Integer
                                                  - Boolean
                                              location: ~
                                              dependencies: ~
                                            inputs:
                                              - kind:
                                                  Expression:
                                                    expression:
                                                      Identifier:
                                                        id: 33
                                                typing: Integer
                                                location: ~
                                                dependencies: ~
                                              - kind:
                                                  Expression:
                                                    expression:
                                                      Identifier:
                                                        id: 29
                                                typing: Integer
                                                location: ~
                                                dependencies: ~
                                    typing: Boolean
                                    location: ~
                                    dependencies: ~
                                  - kind:
                                      Expression:
                                        expression:
                                          Enumeration:
                                            enum_id: 22
                                            elem_id: 20
                                    typing:
                                      Enumeration: ResetState
                                    location: ~
                                    dependencies: ~
                                  - kind:
                                      Expression:
                                        expression:
                                          Enumeration:
                                            enum_id: 22
                                            elem_id: 21
                                    typing:
                                      Enumeration: ResetState
                                    location: ~
                                    dependencies: ~
                        typing:
                          Enumeration: ResetState
                        location: ~
                        dependencies: ~
          typing:
            Enumeration: ResetState
          location: ~
          dependencies: ~
        location: ~
      32:
        id: 32
        expression:
          kind:
            Expression:
              expression:
                Match:
                  expression:
                    kind:
                      Expression:
                        expression:
                          Application:
                            function_expression:
                              kind:
                                Expression:
                                  expression:
                                    Identifier:
                                      id: 2
                              typing:
                                Abstract:
                                  - - Enumeration: Button
                                  - Enumeration: Button
                              location: ~
                              dependencies: ~
                            inputs:
                              - kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 27
                                typing:
                                  Enumeration: Button
                                location: ~
                                dependencies: ~
                    typing:
                      Enumeration: Button
                    location: ~
                    dependencies: ~
                  arms:
                    - - kind:
                          Enumeration:
                            enum_id: 19
                            elem_id: 17
                        typing:
                          Enumeration: Button
                        location: ~
                      - ~
                      - []
                      - kind:
                          FollowedBy:
                            constant:
                              kind:
                                Expression:
                                  expression:
                                    Constant:
                                      constant:
                                        Boolean: false
                              typing: Boolean
                              location: ~
                              dependencies: ~
                            expression:
                              kind:
                                Expression:
                                  expression:
                                    Identifier:
                                      id: 32
                              typing: Boolean
                              location: ~
                              dependencies: ~
                        typing: Boolean
                        location: ~
                        dependencies: ~
                    - - kind:
                          Enumeration:
                            enum_id: 19
                            elem_id: 18
                        typing:
                          Enumeration: Button
                        location: ~
                      - ~
                      - []
                      - kind:
                          Expression:
                            expression:
                              Application:
                                function_expression:
                                  kind:
                                    Expression:
                                      expression:
                                        Identifier:
                                          id: 11
                                  typing:
                                    Abstract:
                                      - - Integer
                                        - Integer
                                      - Boolean
                                  location: ~
                                  dependencies: ~
                                inputs:
                                  - kind:
                                      Expression:
                                        expression:
                                          Identifier:
                                            id: 33
                                    typing: Integer
                                    location: ~
                                    dependencies: ~
                                  - kind:
                                      Expression:
                                        expression:
                                          Application:
                                            function_expression:
                                              kind:
                                                Expression:
                                                  expression:
                                                    Identifier:
                                                      id: 5
                                              typing:
                                                Abstract:
                                                  - - Integer
                                                    - Integer
                                                  - Integer
                                              location: ~
                                              dependencies: ~
                                            inputs:
                                              - kind:
                                                  Expression:
                                                    expression:
                                                      Identifier:
                                                        id: 29
                                                typing: Integer
                                                location: ~
                                                dependencies: ~
                                              - kind:
                                                  Expression:
                                                    expression:
                                                      Identifier:
                                                        id: 30
                                                typing: Integer
                                                location: ~
                                                dependencies: ~
                                    typing: Integer
                                    location: ~
                                    dependencies: ~
                        typing: Boolean
                        location: ~
                        dependencies: ~
          typing: Boolean
          location: ~
          dependencies: ~
        location: ~
      33:
        id: 33
        expression:
          kind:
            NodeApplication:
              node_id: 26
              inputs:
                - - 23
                  - kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 34
                    typing: Boolean
                    location: ~
                    dependencies: ~
                - - 24
                  - kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 28
                    typing: Integer
                    location: ~
                    dependencies: ~
              output_id: 25
          typing: Integer
          location: ~
          dependencies: ~
        location: ~
      34:
        id: 34
        expression:
          kind:
            Expression:
              expression:
                Application:
                  function_expression:
                    kind:
                      Expression:
                        expression:
                          Identifier:
                            id: 7
                    typing:
                      Abstract:
                        - - Boolean
                          - Boolean
                        - Boolean
                    location: ~
                    dependencies: ~
                  inputs:
                    - kind:
                        Expression:
                          expression:
                            Application:
                              function_expression:
                                kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 2
                                typing:
                                  Abstract:
                                    - - Boolean
                                    - Boolean
                                location: ~
                                dependencies: ~
                              inputs:
                                - kind:
                                    Expression:
                                      expression:
                                        Application:
                                          function_expression:
                                            kind:
                                              Expression:
                                                expression:
                                                  Identifier:
                                                    id: 9
                                            typing:
                                              Abstract:
                                                - - Enumeration: Button
                                                  - Enumeration: Button
                                                - Boolean
                                            location: ~
                                            dependencies: ~
                                          inputs:
                                            - kind:
                                                Expression:
                                                  expression:
                                                    Identifier:
                                                      id: 27
                                              typing:
                                                Enumeration: Button
                                              location: ~
                                              dependencies: ~
                                            - kind:
                                                Expression:
                                                  expression:
                                                    Enumeration:
                                                      enum_id: 19
                                                      elem_id: 18
                                              typing:
                                                Enumeration: Button
                                              location: ~
                                              dependencies: ~
                                  typing: Boolean
                                  location: ~
                                  dependencies: ~
                      typing: Boolean
                      location: ~
                      dependencies: ~
                    - kind:
                        Expression:
                          expression:
                            Application:
                              function_expression:
                                kind:
                                  Expression:
                                    expression:
                                      Identifier:
                                        id: 2
                                typing:
                                  Abstract:
                                    - - Boolean
                                    - Boolean
                                location: ~
                                dependencies: ~
                              inputs:
                                - kind:
                                    FollowedBy:
                                      constant:
                                        kind:
                                          Expression:
                                            expression:
                                              Constant:
                                                constant:
                                                  Boolean: true
                                        typing: Boolean
                                        location: ~
                                        dependencies: ~
                                      expression:
                                        kind:
                                          Expression:
                                            expression:
                                              Application:
                                                function_expression:
                                                  kind:
                                                    Expression:
                                                      expression:
                                                        Identifier:
                                                          id: 9
                                                  typing:
                                                    Abstract:
                                                      - - Enumeration: Button
                                                        - Enumeration: Button
                                                      - Boolean
                                                  location: ~
                                                  dependencies: ~
                                                inputs:
                                                  - kind:
                                                      Expression:
                                                        expression:
                                                          Identifier:
                                                            id: 27
                                                    typing:
                                                      Enumeration: Button
                                                    location: ~
                                                    dependencies: ~
                                                  - kind:
                                                      Expression:
                                                        expression:
                                                          Enumeration:
                                                            enum_id: 19
                                                            elem_id: 17
                                                    typing:
                                                      Enumeration: Button
                                                    location: ~
                                                    dependencies: ~
                                        typing: Boolean
                                        location: ~
                                        dependencies: ~
                                  typing: Boolean
                                  location: ~
                                  dependencies: ~
                      typing: Boolean
                      location: ~
                      dependencies: ~
          typing: Boolean
          location: ~
          dependencies: ~
        location: ~
    unitary_nodes: {}
    contract:
      requires: []
      ensures: []
      invariant: []
    location: ~
    graph: ~
component: ~
location: ~
